File: solutions/ruby/yacht/2/yacht.rb
class Yacht
  attr_reader :score

  def initialize(rolls, category)
    @score = 0

    case category
    when 'ones'
      @score = rolls.count(1)
    when 'twos'
      @score = rolls.count(2) * 2
    when 'threes'
      @score = rolls.count(3) * 3
    when 'fours'
      @score = rolls.count(4) * 4
    when 'fives'
      @score = rolls.count(5) * 4
    when 'sixes'
      @score = rolls.count(6) * 6
    when 'full house'
      @score = rolls.sum if full_house?(rolls)
    when 'four of a kind'
      @score = four_of_a_kind(rolls)
    when 'little straight'
      @score = 30 if rolls.sort == [1, 2, 3, 4, 5]
    when 'big straight'
      @score = 30 if rolls.sort == [2, 3, 4, 5, 6]
    when 'choice'
      @score = rolls.sum
    when 'yacht'
      @score = 50 if rolls.all? { |r| r == rolls.first }
    end
  end

  private

  def full_house?(rolls)
    unique_counts = rolls.uniq.map { |r| rolls.count(r) }
    unique_counts.sort == [2, 3]
  end

  def four_of_a_kind(rolls)
    sorted      = rolls.sort
    first_count = rolls.count(sorted.first)
    last_count  = rolls.count(sorted.last)

    return sorted.first * 4 if first_count >= 4
    return sorted.last  * 4 if last_count  >= 4
    return 0
  end
end

File: solutions/javascript/yacht/2/yacht.js
const scoreFullHouse = (rolls) => {
  if (rolls.length !== 5) return 0;

  const sorted = rolls.sort();

  if (sorted[0] !== sorted[4]
  &&  sorted[1] === sorted[0]
  &&  sorted[3] === sorted[4]
  && (sorted[2] === sorted[0] || sorted[2] === sorted[4])) {
    return rolls.reduce((sum, roll) => sum += roll);
  }

  return 0;
};

const scoreFourOfAKind = (rolls) => {
  if (rolls.length !== 5) return 0;

  const sorted     = rolls.sort();
  const first      = sorted[0];
  const last       = sorted[4];
  const first_four = sorted.slice(0, 4);
  const last_four  = sorted.slice(1, 5);

  if (first_four.every(r => r === first)) {
    return first * 4;
  }

  if (last_four.every(r => r === last)) {
    return last * 4;
  }

  return 0;
};

export const score = (rolls, category) => {
  switch (category) {
    case 'ones':
      return rolls.filter(r => r === 1).reduce((sum, roll) => sum += roll, 0);
    case 'twos':
      return rolls.filter(r => r === 2).reduce((sum, roll) => sum += roll, 0);
    case 'threes':
      return rolls.filter(r => r === 3).reduce((sum, roll) => sum += roll, 0);
    case 'fours':
      return rolls.filter(r => r === 4).reduce((sum, roll) => sum += roll, 0);
    case 'fives':
      return rolls.filter(r => r === 5).reduce((sum, roll) => sum += roll, 0);
    case 'sixes':
      return rolls.filter(r => r === 6).reduce((sum, roll) => sum += roll, 0);
    case 'full house':
      return scoreFullHouse(rolls);
    case 'four of a kind':
      return scoreFourOfAKind(rolls);
    case 'little straight':
      const little = [1, 2, 3, 4, 5];
      return rolls.sort().every((roll, i) => roll === little[i])  ? 30 : 0;
    case 'big straight':
      const big = [2, 3, 4, 5, 6];
      return rolls.sort().every((roll, i) => roll === big[i])  ? 30 : 0;
    case 'choice':
      return rolls.reduce((sum, roll) => sum += roll);
    case 'yacht':
      return rolls.every(r => r === rolls[0]) ? 50 : 0;
    default:
      return 0;
  }
};
